utils/rowgroup/rowaggregation.cpp

```diff
@@ -2666,14 +2703,38 @@ void RowAggregationUM::calculateAvgColumns()
                 if (cnt == 0) // empty set, value is initialized to null.
                     continue;
 
-                long double sum = 0.0;
-                long double avg = 0.0;
+                uint32_t precision = fRow.getPrecision(colOut);
+                bool isWideDecimal =
+                    datatypes::Decimal::isWideDecimalType(precision);
 
-                // MCOL-1822 Always long double
-                sum = fRow.getLongDoubleField(colOut);
-                avg = sum / cnt;
-//                avg *= factor;
-                fRow.setLongDoubleField(avg, colOut);
+                if (LIKELY(!isWideDecimal))
+                {
+                    long double sum = 0.0;
+                    long double avg = 0.0;
+                    sum = fRow.getLongDoubleField(colOut);
+                    avg = sum / cnt;
+                    fRow.setLongDoubleField(avg, colOut);
+                }
+                else
+                {
+                    uint32_t offset = fRow.getOffset(colOut);
+                    uint32_t scale = fRow.getScale(colOut);
+                    // Get multiplied to deliver AVG with the scale closest
+                    // to the expected original scale + 4. 
+                    // There is a counterpart in buildAggregateColumn.
+                    datatypes::Decimal::setScalePrecision4Avg(precision, scale);
+                    int128_t* sumPnt = fRow.getBinaryField_offset<int128_t>(offset);
+                    uint32_t scaleDiff = scale - fRow.getScale(colOut);
+                    // multiplication overflow check
+                    datatypes::MultiplicationOverflowCheck multOp;
+                    int128_t sum = 0;
+                    if (scaleDiff > 0)
+                        multOp(*sumPnt, datatypes::mcs_pow_10[scaleDiff], sum);
+                    else
+                        sum = *sumPnt;
+                    int128_t avg = sum / cnt;
```

Comment:
I think we can have a rounding error here, e.g. if sum=14, cnt=5, avg=2.8. This should be rounded up to 3, and not 2. We can refer to the divisionExecute() code in mcs_decimal.cpp for similar rounding handling.

Reply:
That implementation has its drawbacks, namely float128_t looses its precision for some numbers that utilizes most of 38 digits available. With the rounding applied avg(38)_test1 returns 0.01122334445566778844556666161821143552 for the avg(d3) column whilst MDB and algo w/o rounding both return 0.01122334445566778844556666161821143636. The former has a higher measure of inaccuracy if we multiply the calculated average to get the original sum.
This is somehow expected given that float must have a space for its exponent. I'll try to come up with a solution or resort to divisionExecute() approach for math.